include/open62541/plugin/pki.h

```diff
@@ -30,6 +30,29 @@ _UA_BEGIN_DECLS
  * The lifecycle of the plugin is attached to a server or client config. The
  * ``clear`` method is called automatically when the config is destroyed. */
 
+/* Enum and functions for the private key password */
+typedef enum  {
+    UA_PRIVATEKEYPASSWORDSTATE_INITIAL = 0,
+    UA_PRIVATEKEYPASSWORDSTATE_WRONG = 1,
+} UA_PrivateKeyPasswordState;
+
+/* This callback is called when a client with security is initialized with a password
+ * protected private key in the PEM format.
+ * If a wrong password is encountered, the callback will be called repeatedly until
+ * doRetry is set to false. */
+typedef UA_String (*UA_PrivateKeyPasswordCallback)(UA_PrivateKeyPasswordState state,
+                                                   UA_Boolean *doRetry, void *context);
+
+/* Attempts to load privateKey and writes a DER encoded private key to derOutput.
+ * If the key is PEM encoded and password protected, the password callback
+ * is called to get the password to decrypt the private key.
+ * A DER private key input is just copied without any loading operation involved.
+ * The content of derOutput must be cleared after use. */
+UA_EXPORT UA_Boolean
+UA_PrivateKey_decryptPemWithPassword(UA_ByteString privateKey, UA_ByteString *derOutput,
```

Comment:
This header is for API definitions for exchangeable plugins.
Right now we do a hard linking of this `UA_EXPORT` method against either mbedTLS or OpenSSL.
Please move the method to one of the headers in the /plugins/include folder.

The generic stuff can stay here.

Reply:
@jpfr Which header would you suggest? I didn't find one that looks fitting.